### PR TITLE
Fix for rebuilding the gage when it leaves the viewport or is hidden/reshown

### PR DESCRIPTION
**Pull request by dnfield**
Adds method to check whether the element is in the viewport - if so, rebuild the justgage to ensure that it animates/displays correctly.  Otherwise, the gage will sometimes not fully redraw.

### CODE REVIEW

File: src/directives/justgage.js [hunk 1]
```diff
@@ -85,6 +102,14 @@ angular.module('frapontillo.gage.directives', ['frapontillo.gage.controllers'])
                 justgage.refresh(scope.value, newValue);
               }
             }));
+            // will fire when the justgage element is brought back into view
+            // if something slides it off the screen or sets ng-show/hide this will ensure the justgage is redrawn and animated properly
+            watchers.push(scope.$watch(function () { return isElementInViewPort(element[0]); }, function (newValue) {
```
Comment (frapontillo) [inline src/directives/justgage.js]: I have to say I'm not a fan of this approach: recalculating a function at every digest is too expensive, but I also know there are no other alternatives for this.

Comment (dnfield) [reply src/directives/justgage.js]: I agree it's not ideal - I couldn't come up with a better way short of getting deeper into the justgage code and reworking how the SVG is animated.  Maybe add a property to allow this watch to be disabled?

Comment (frapontillo) [reply src/directives/justgage.js]: YES PLEASE

Comment (dnfield) [reply src/directives/justgage.js]: Updated - new attribute, watch-view-port, set to true to set up that watcher, false to not.  It wouldn't response to changes, not 100% sure how I'd make that work

File: src/directives/justgage.js [hunk 2]
```diff
@@ -64,13 +64,31 @@ angular.module('frapontillo.gage.directives', ['frapontillo.gage.controllers'])
...
+          textRenderer: '&',                 // function applied before rendering text
+          watchViewPort: '@'                 // Watch for the element going out of the viewport/being hidden - redraw to prevent UI issues
         },
         controller: 'justgageCtrl',
         link: function (scope, element, attrs, justgageCtrl) {
           var justgage,
               watchers = [];
 
+          // used to determine whether the justgage element is in the viewport or not.
+          function isElementInViewPort(el) {
```
Comment (frapontillo) [inline src/directives/justgage.js]: You already have `element`, don't need an argument here at all.

File: src/directives/justgage.js [hunk 3]
```diff
@@ -64,13 +64,31 @@ angular.module('frapontillo.gage.directives', ['frapontillo.gage.controllers'])
...
+            return (
+              rect.top >= 0 &&
+              rect.left >= 0 &&
+              rect.bottom <= (window.innerHeight || document.documentElement.clientHeight) &&
+              rect.right <= (window.innerWidth || document.documentElement.clientWidth)
+            );
+          }
+          
+          // get rid of the justgage when we're done
+          scope.$on('$destroy', function handleDestroyEvent() {
```
Comment (frapontillo) [inline src/directives/justgage.js]: You can pass the `destroyJustgage` function as an argument.

File: src/directives/justgage.js [hunk 4]
```diff
@@ -85,6 +103,17 @@ angular.module('frapontillo.gage.directives', ['frapontillo.gage.controllers'])
                 justgage.refresh(scope.value, newValue);
               }
             }));
+            
+            // will fire only if the attribute 'watch-view-port' is set to true
+            // if fired, justgage element is brought back into view
+            // if something slides it off the screen or sets ng-show/hide this will ensure the justgage is redrawn and animated properly
+            if (justgageCtrl.getOptionValue('watchViewPort', scope.watchViewPort) === true) {
+              watchers.push(scope.$watch(function () { return isElementInViewPort(element[0]); }, function (newValue) {
```
Comment (frapontillo) [inline src/directives/justgage.js]: By removing the `el` argument to the function `isElementInViewPort` you can pass it as a simple argument to `$watch`.

File: src/directives/justgage.js [hunk 5]
```diff
@@ -102,31 +131,36 @@ angular.module('frapontillo.gage.directives', ['frapontillo.gage.controllers'])
...
+          // helper functions to destroy/rebuild justgage
+          // used outside of init() now, since we have to do this when the gage is brought back into view
+          var destroyJustgage = function() {
+            if (justgage) {
+              var canvasDom = justgage.canvas.canvas;
+              canvasDom.parentNode.removeChild(canvasDom);
+              justgage = null;
+            }
+          };
+          var rebuildJustgage = function() {
```
Comment (frapontillo) [inline src/directives/justgage.js]: Please add a blank line before the `rebuildJustgage` function.

**Review by frapontillo — CHANGES_REQUESTED** [review]
A few improvements, then good to go.

**Comment by dnfield** [comment]
Done, thanks for the feedback.

**Comment by dnfield** [comment]
Not sure if this project is still alive or anyone's interested - feel free to reopen and merge if you want, I'm just trying to clean up what shows up in my open pulls list :)

**Pull request closed without merging**